# Assignment 3 - Text Analysis
An explanation this assignment could be found in the .pdf explanation document


## Materials to review for this assignment
<h4>From Moodle:</h4> 
<h5><u>Review the notebooks regarding the following python topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Working with strings</b> (tutorial notebook)<br/>
&#x2714; <b>Text Analysis</b> (tutorial notebook)<br/>
&#x2714; <b>Hebrew text analysis tools (tokenizer, wordnet)</b> (moodle example)<br/>
&#x2714; <b>(brief review) All previous notebooks</b><br/>
</div> 
<h5><u>Review the presentations regarding the following topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Text Analysis</b> (lecture presentation)<br/>
&#x2714; <b>(brief review) All other presentations</b><br/>
</div>

## Personal Details:

In [1]:
# Details Student 1:

# Details Student 2:


## Preceding Step - import modules (packages)
This step is necessary in order to use external modules (packages). <br/>

In [2]:
# --------------------------------------
import pandas as pd
import numpy as np
# --------------------------------------


# --------------------------------------
# ------------- visualizations:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
# --------------------------------------


# ---------------------------------------
import sklearn
from sklearn import preprocessing, metrics, pipeline, model_selection, feature_extraction 
from sklearn import naive_bayes, linear_model, svm, neural_network, neighbors, tree
from sklearn import decomposition, cluster

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV 
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report
from sklearn.metrics import mean_squared_error, r2_score, silhouette_score
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder

from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import Perceptron, SGDClassifier
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# ---------------------------------------


# ----------------- output and visualizations: 
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.simplefilter("ignore")
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)
# show several prints in one cell. This will allow us to condence every trick in one cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%matplotlib inline
pd.pandas.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# ---------------------------------------

### Text analysis and String manipulation imports:

In [3]:
# --------------------------------------
# --------- Text analysis and Hebrew text analysis imports:
# vectorizers:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# regular expressions:
import re
# --------------------------------------

### (optional) Hebrew text analysis - WordNet (for Hebrew)
Note: the WordNet is not a must

#### (optional) Only if you didn't install Wordnet (for Hebrew) use:

In [4]:
# word net installation:

# unmark if you want to use and need to install
# !pip install wn
# !python -m wn download omw-he:1.4

In [5]:
# word net import:

# unmark if you want to use:
# import wn

### (optional) Hebrew text analysis - hebrew_tokenizer (Tokenizer for Hebrew)
Note: the hebrew_tokenizer is not a must

#### (optional) Only if you didn't install hebrew_tokenizer use:

In [6]:
# Hebrew tokenizer installation:

# unmark if you want to use and need to install:
# !pip install hebrew_tokenizer

In [7]:
# Hebrew tokenizer import:

# unmark if you want to use:
import hebrew_tokenizer as ht

C:\Users\aytan\Desktop\למידת מכונה\משימה 3\assignment3-text_analysis


### Reading input files
Reading input files for train annotated corpus (raw text data) corpus and for the test corpus

In [8]:
train_filename = 'annotated_corpus_for_train.csv'
test_filename  = 'corpus_for_test.csv'
df_train = pd.read_csv(train_filename, index_col=None, encoding='utf-8')
df_test  = pd.read_csv(test_filename, index_col=None, encoding='utf-8')

In [9]:
df_train.head(8)
df_train.shape

story gender
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...      m
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...      m
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...      f
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...      m
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...      f
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...      f
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...      m
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...      f

(753, 2)

In [10]:
df_test.head(3)
df_test.shape

test_example_id                                              story
0                0  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                1  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                2  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...

(323, 2)

### Your implementation:
Write your code solution in the following code-cells

In [11]:
# create the numeric label and visualize in one table with categorical lables (m=0, f=1)
y = pd.Series(df_train.gender.factorize()[0])
df_y = pd.concat([y, df_train.gender], axis=1)
df_y

0 gender
0    0      m
1    0      m
2    1      f
3    0      m
4    1      f
..  ..    ...
748  0      m
749  0      m
750  0      m
751  1      f
752  0      m

[753 rows x 2 columns]

# first try with fold out for self understending and general methods build

In [12]:
# split train corpus for train and test
# the story col is the corpus, the y is the factorized categories (m=0, f=1)
X_train, X_test, y_train, y_test = train_test_split(
    df_train.story, y, test_size=0.20, random_state=42)

In [13]:
# YOUR CODE HERE

# tokenize story
def story_tokanize(story):
    tok_stor = ht.tokenize(story)
    tok_stor_list = []
    for tok in tok_stor:
        # remove all the the non
        if(tok[0] != 'HEBREW'):
            continue
        tok_stor_list.append(tok[1])
    tok_stor = [" ".join(tokens) for tokens in tok_stor_list]
    return tok_stor

In [14]:
# after first vectorization with to much columns, i decided to nurrow to 10k features only 
# (originaly more then 20k with 1gram and more then 90k with 1grams)

vectorizer = CountVectorizer(ngram_range=(1,1), max_features=10000, tokenizer=story_tokanize)

In [15]:
# choosing the best parameters for the classifier
from sklearn.metrics import make_scorer
clf_dt = GridSearchCV(estimator=DecisionTreeClassifier(),
                      param_grid={'max_depth':[2,3,4,5,6,8,10],
                                  "min_samples_split":[3,5,8,10,15,20],
                                  'min_samples_leaf':[1,2,3,4,5]},
                      scoring=make_scorer(metrics.f1_score, greater_is_better=True),
                      n_jobs=-1)

In [16]:
# define pipeline for decision tree
text_clf = Pipeline([
    ('vect', vectorizer), # the vectorizer defined above
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_dt', clf_dt)
])
text_clf.fit(X_train, y_train)
predicted = text_clf.predict(X_test)
print(f'macro avarage is: {f1_score(y_true=y_test, y_pred=predicted, average="macro")}')

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=10000,
                                 tokenizer=<function story_tokanize at 0x0000016E42657430>)),
                ('norm', Normalizer(norm='l1')),
                ('clf_dt',
                 GridSearchCV(estimator=DecisionTreeClassifier(), n_jobs=-1,
                              param_grid={'max_depth': [2, 3, 4, 5, 6, 8, 10],
                                          'min_samples_leaf': [1, 2, 3, 4, 5],
                                          'min_samples_split': [3, 5, 8, 10, 15,
                                                                20]},
                              scoring=make_scorer(f1_score)))])

macro avarage is: 0.6599099099099099


In [17]:
# test only to narrow the grid search
clf_dt = GridSearchCV(estimator=DecisionTreeClassifier(),
                      param_grid={'max_depth':[2,4,6,10],
                                  "min_samples_split":[3,6,9,12,16],
                                  'min_samples_leaf':[1,3,5]},
                      scoring=make_scorer(metrics.f1_score, greater_is_better=True),
                      n_jobs=-1)
# define pipeline for decision tree
text_clf = Pipeline([
    ('vect', vectorizer), # the vectorizer defined above
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_dt', clf_dt)
])
text_clf.fit(X_train, y_train)
predicted = text_clf.predict(X_test)
print(f'macro avarage is: {f1_score(y_true=y_test, y_pred=predicted, average="macro")}')

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=10000,
                                 tokenizer=<function story_tokanize at 0x0000016E42657430>)),
                ('norm', Normalizer(norm='l1')),
                ('clf_dt',
                 GridSearchCV(estimator=DecisionTreeClassifier(), n_jobs=-1,
                              param_grid={'max_depth': [2, 4, 6, 10],
                                          'min_samples_leaf': [1, 3, 5],
                                          'min_samples_split': [3, 6, 9, 12,
                                                                16]},
                              scoring=make_scorer(f1_score)))])

macro avarage is: 0.6466748768472906


# Estimating with 10 fold cross validation

Trying different models and hyperparameters.
Highest average (of all folds) macro f1 score of all model will determine wich model is better.

In [18]:
# calc avg f1 = (f1_male + f1_female)/2 and present it
def model_estimation(text_clf, X, y):
    cv_arr = cross_val_score(text_clf, X, y, cv=10, n_jobs=-1, scoring='f1_macro')
    print('array of macro avarages of the model is:')
    print(cv_arr)
    print(f'mean of all macro f_score is: {cv_arr.mean()}')

# Decision Tree Model

In [19]:
############################ cross validation ###################################

# y is the class series (m=0,f=1)
# df_train.story is the X
X = df_train.story

# desition tree (grid searched parameters)
# CountVectorizer vectorization with 1gram
# hebrew_tokanization and saving hebrow only
model_estimation(text_clf, X, y)

array of macro avarages of the model is:
[0.67099567 0.57451346 0.703125   0.58333333 0.65277778 0.52574526
 0.54166667 0.58333333 0.69503931 0.47306792]
mean of all macro f_score is: 0.6003597724371734


In [21]:
# define pipeline for decision tree - ngram changed to (1,2)gram
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_dt', clf_dt)
])
model_estimation(text_clf, X, y)

array of macro avarages of the model is:
[0.54519685 0.52167832 0.60416667 0.50396825 0.62779156 0.50840695
 0.4882134  0.496      0.592      0.64247881]
mean of all macro f_score is: 0.5529900822593441


In [22]:
# define pipeline for decision tree - ngram changed to (1,2)gram, vectorization canged to tfidf
text_clf = Pipeline([
    ('vect', TfidfVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_dt', clf_dt)
])
model_estimation(text_clf, X, y)

array of macro avarages of the model is:
[0.56221198 0.52711111 0.56454331 0.58126551 0.67939584 0.58333333
 0.60480094 0.544      0.60480094 0.50396825]
mean of all macro f_score is: 0.5755431211211609


Decision tree did not gave us the best prediction.

lets try another model that works with high dimentionals feature vectors.

# SVC Model

In [23]:
# define first pipeline for svc for reference before tuning
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_svc', LinearSVC())
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.43283582 0.43283582 0.43283582 0.43609023 0.43609023 0.43181818
 0.43181818 0.43181818 0.43181818 0.43181818]
mean of all macro f_score is: 0.4329778822905296


In [24]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l1 (manhattan) normalizer
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.65948368 0.57212192 0.74400634 0.581698   0.68322981 0.63450292
 0.73015873 0.55681818 0.60171569 0.6015246 ]
mean of all macro f_score is: 0.6365259870432399


In [25]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l1 (manhattan) normalizer
# 1,2gram word vectorization
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.65948368 0.5827451  0.73176471 0.56103679 0.66977467 0.64596273
 0.71780936 0.5468785  0.63235294 0.6015246 ]
mean of all macro f_score is: 0.6349333077729633


In [26]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# 1,2gram word vectorization
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l1')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.65948368 0.5827451  0.73176471 0.56103679 0.66977467 0.64596273
 0.71780936 0.5468785  0.63235294 0.6015246 ]
mean of all macro f_score is: 0.6349333077729633


In [27]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1gram only
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.70881226 0.5621933  0.83296703 0.65770791 0.68220339 0.65770791
 0.7668022  0.66124661 0.72049689 0.76165254]
mean of all macro f_score is: 0.7011790055133149


In [28]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1,2gram only
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.73479561 0.58250934 0.77268195 0.65437788 0.68220339 0.65770791
 0.73092562 0.66124661 0.72192797 0.79166667]
mean of all macro f_score is: 0.6990042947786208


In [29]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1,2,3gram only
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,3), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.73479561 0.63809524 0.77268195 0.65437788 0.68220339 0.65770791
 0.74415205 0.66124661 0.74747475 0.79166667]
mean of all macro f_score is: 0.7084402059549856


In [30]:
# define pipeline for svc with ballanced class wight hyperparameter
# using count vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1,2,3gram only
# got max feachers up to 15k
print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,3), max_features=15000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.73479561 0.59285714 0.78711485 0.65437788 0.69503931 0.68220339
 0.77180527 0.66124661 0.74747475 0.76489028]
mean of all macro f_score is: 0.7091805093922081


so far the best f1 was for the cell above 70.9%

lets try tfidf vectorization

In [31]:
# define pipeline for svc with ballanced class wight hyperparameter
# using tfidf vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1,2,3gram only
# got max feachers up to 15k
print('clf_svc:')
text_clf = Pipeline([
    ('vect', TfidfVectorizer(ngram_range=(1,3), max_features=15000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.69354839 0.70928962 0.61577351 0.73653396 0.65277778 0.61607949
 0.67431762 0.6031746  0.72084367 0.62713068]
mean of all macro f_score is: 0.6649469318244504


In [32]:
# define pipeline for svc with ballanced class wight hyperparameter
# using tfidf vectorizer
# tokenizer implemented using hebrew_tokenizer
# using l2 (euclidean) normalizer
# with 1,2,3gram only
# got max feachers down to 10k
print('clf_svc:')
text_clf = Pipeline([
    ('vect', TfidfVectorizer(ngram_range=(1,3), max_features=10000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.69354839 0.70928962 0.70725926 0.72214306 0.67431762 0.64871194
 0.70641373 0.64       0.70641373 0.70238095]
mean of all macro f_score is: 0.6910478300192244


not to bad but the last count vectorizer was a bit better.

going back to the best pipeline combination found and try to tune parameters.

In [33]:
# C = 0.1 (defoult 1)
# using the best performance found before:
    # define pipeline for svc with ballanced class wight hyperparameter
    # using count vectorizer
    # tokenizer implemented using hebrew_tokenizer
    # using l2 (euclidean) normalizer
    # with 1,2,3gram only
    # got max feachers up to 15k

print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,3), max_features=15000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced', C=0.1))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.68421053 0.57371795 0.78290486 0.63624339 0.68220339 0.65917245
 0.72288633 0.581643   0.71054889 0.65437788]
mean of all macro f_score is: 0.6687908650315103


In [34]:
# C = 10 (defoult 1)
# using the best performance found before:
    # define pipeline for svc with ballanced class wight hyperparameter
    # using count vectorizer
    # tokenizer implemented using hebrew_tokenizer
    # using l2 (euclidean) normalizer
    # with 1,2,3gram only
    # got max feachers up to 15k

print('clf_svc:')
text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,3), max_features=15000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced', C=10))
])
model_estimation(text_clf, X, y)

clf_svc:
array of macro avarages of the model is:
[0.68333333 0.56458333 0.69354839 0.67939584 0.64       0.65437788
 0.69503931 0.69262295 0.66666667 0.72084367]
mean of all macro f_score is: 0.6690411369206133


# Final classifier we got is the following with the best macro f1 score:

define pipeline for svc with ballanced class wight hyperparameter

using count vectorizer

tokenizer implemented using hebrew_tokenizer

using l2 (euclidean) normalizer

with 1,2,3gram only

got max feachers up to 15k

In [35]:
final_text_clf = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,3), max_features=15000, tokenizer=story_tokanize)),
    ('norm', preprocessing.Normalizer(norm='l2')),
    ('clf_svc', LinearSVC(class_weight='balanced'))
])
model_estimation(final_text_clf, X, y)

array of macro avarages of the model is:
[0.73479561 0.59285714 0.78711485 0.65437788 0.69503931 0.68220339
 0.77180527 0.66124661 0.74747475 0.76489028]
mean of all macro f_score is: 0.7091805093922081


# final check of the classifier with hold out

just to be sure*

In [36]:
X_train, X_test, y_train, y_test = train_test_split(
    df_train.story, y, test_size=0.20, random_state=42)
final_text_clf.fit(X_train,y_train)

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=15000, ngram_range=(1, 3),
                                 tokenizer=<function story_tokanize at 0x0000016E42657430>)),
                ('norm', Normalizer()),
                ('clf_svc', LinearSVC(class_weight='balanced'))])

In [37]:
y_pred = final_text_clf.predict(X_test)

In [38]:
f1_score(y_true=y_test, y_pred=y_pred, average='macro')

0.6773504273504274

#  Training of the Final model with the train set

In [39]:
# fit
final_text_clf.fit(X,y)
# predict - 0=m , 1=f
pred_test_0_1 = final_text_clf.predict(df_test.story)
pred_test_0_1

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=15000, ngram_range=(1, 3),
                                 tokenizer=<function story_tokanize at 0x0000016E42657430>)),
                ('norm', Normalizer()),
                ('clf_svc', LinearSVC(class_weight='balanced'))])

array([0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0,

In [40]:
# transforming the binary array to m/f series.
pred_test = pd.Series(pred_test_0_1, name='predicted_category').replace({0:'m',1:'f'})
pred_test

0      m
1      m
2      m
3      m
4      f
      ..
318    m
319    m
320    m
321    m
322    m
Name: predicted_category, Length: 323, dtype: object

In [41]:
# concatenation of the original indexes from test data to the results as requested.
df_predicted = pd.concat([df_test.test_example_id, pred_test], axis=1)
# show 5 first and last predictions.
df_predicted

test_example_id predicted_category
0                  0                  m
1                  1                  m
2                  2                  m
3                  3                  m
4                  4                  f
..               ...                ...
318              318                  m
319              319                  m
320              320                  m
321              321                  m
322              322                  m

[323 rows x 2 columns]

### Save output to csv (optional)
After you're done save your output to the 'classification_results.csv' csv file.<br/>
We assume that the dataframe with your results contain the following columns:
* column 1 (left column): 'test_example_id'  - the same id associated to each of the test stories to be predicted.
* column 2 (right column): 'predicted_category' - the predicted gender value for each of the associated story. 

Assuming your predicted values are in the `df_predicted` dataframe, you should save you're results as following:

In [42]:
df_predicted.to_csv('classification_results.csv',index=False)